File: parser.go
package main

import (
	"github.com/paidgeek/ssdl/token"
	"github.com/paidgeek/ssdl/ast"
)

type Parser struct {
	lexer   *Lexer
	current Symbol
	next    Symbol
}

func (p *Parser) Parse() *ast.Document {
	p.next = p.lexer.Next()
	definitions := make([]ast.Definition, 0)

	for !p.match(token.EOF) {
		definitions = append(definitions, p.parseDefinition())
	}

	return &ast.Document{
		Definitions:definitions,
	}
}

func (p *Parser) parseDefinition() ast.Definition {
	switch p.peek() {
	case token.KeywordType:
		return p.parseTypeDefinition()
	case token.KeywordEnum:
		return p.parseEnumDefinition()
	}

	panic(reportf(p.current.Position, "unexpected '%v'", p.next.Lexeme))
	return nil
}

func (p *Parser) parseTypeDefinition() *ast.TypeDefinition {
	p.check(token.KeywordType)
	name := p.parseIdentifier()
	fields := p.parseFieldList()

	return &ast.TypeDefinition{
		Name:name,
		Fields:fields,
	}
}

func (p *Parser) parseEnumDefinition() *ast.EnumDefinition {
	p.check(token.KeywordEnum)
	name := p.parseIdentifier()
	values := p.parseIdentifierList()

	return &ast.EnumDefinition{
		Name:name,
		Values:values,
	}
}

func (p *Parser) parseFieldList() []*ast.Field {
	p.check(token.OpenBrace)
	fields := make([]*ast.Field, 0)

	for p.match(token.Identifier) {
		fields = append(fields, p.parseField())
		p.accept(token.Comma)
	}
	p.check(token.CloseBrace)

	return fields
}

func (p *Parser) parseIdentifierList() []*ast.Identifier {
	p.check(token.OpenBrace)
	identifiers := make([]*ast.Identifier, 0)

	for p.match(token.Identifier) {
		identifiers = append(identifiers, p.parseIdentifier())
		p.accept(token.Comma)
	}
	p.check(token.CloseBrace)

	return identifiers
}

func (p *Parser) parseField() *ast.Field {
	name := p.parseIdentifier()
	t := p.parseType()

	return &ast.Field{
		Name:name,
		Type:t,
	}
}

func (p *Parser) parseIdentifier() *ast.Identifier {
	p.check(token.Identifier)
	return &ast.Identifier{
		Name:p.current.Lexeme,
	}
}

func (p *Parser) parseType() *ast.Type {
	isList := p.accept(token.Brackets)

	if p.match(token.OpenBrace) {
		fields := p.parseFieldList()

		return &ast.Type{
			IsInnerType:true,
			Fields:fields,
		}
	}

	isReference := p.accept(token.Asterisk)

	if p.match(token.Identifier) {
		return &ast.Type{
			Name:p.parseIdentifier(),
			Token:token.Identifier,
			IsList:isList,
			IsReference:isReference,
		}
	} else {
		if !p.peek().IsType() {
			panic(reportf(p.current.Position, "expected type, got '%v'", p.next.Token))
		}

		p.accept(p.peek())

		return &ast.Type{
			Token:p.current.Token,
			IsList:isList,
			IsReference:isReference,
		}
	}
}

func (p *Parser) peek() token.Token {
	return p.next.Token
}

func (p *Parser) match(token token.Token) bool {
	return p.peek() == token
}

func (p *Parser) check(tokens ...token.Token) {
	for _, t := range tokens {
		if p.match(t) {
			p.current = p.next
			p.next = p.lexer.Next()
		} else {
			panic(reportf(p.current.Position, "expected '%v', got '%v'", t, p.next.Token))
		}
	}
}

func (p *Parser) accept(tokens ...token.Token) bool {
	for _, t := range tokens {
		if p.match(t) {
			p.current = p.next
			p.next = p.lexer.Next()
			return true
		}
	}
	return false
}

func NewParser(lexer *Lexer) *Parser {
	return &Parser{
		lexer:lexer,
	}
}

File: src/main/java/com/moybl/ssdl/Parser.java
package com.moybl.ssdl;

import com.moybl.ssdl.ast.*;

import java.util.ArrayList;
import java.util.List;

public class Parser {

	private Lexer lexer;
	private Symbol current;
	private Symbol next;

	public Parser(Lexer lexer) {
		this.lexer = lexer;
	}

	public Schema parse() {
		next = lexer.next();

		List<Definition> definitions = parseDefinitions();
		Position p = Position
				.expand(definitions.get(0).getPosition(), definitions.get(definitions.size() - 1)
						.getPosition());

		return new Schema(p, definitions);
	}

	private List<Definition> parseDefinitions() {
		List<Definition> definitions = new ArrayList<Definition>();

		while (!match(Token.EOF)) {
			Definition definition = parseDefinition();

			if (definition != null) {
				definitions.add(definition);
			}
		}

		return definitions;
	}

	private Definition parseDefinition() {
		switch (next.getToken()) {
			case KEYWORD_TYPE:
				return parseTypeDefinition();
			case KEYWORD_ENUM:
				return parseEnumDefinition();
		}

		throw ParserException.internal();
	}

	private TypeDefinition parseTypeDefinition() {
		check(Token.KEYWORD_TYPE);
		Position a = current.getPosition();
		Identifier name = parseIdentifier();

		if (match(Token.IDENTIFIER)) {
			Identifier oldName = parseIdentifier();

			return new TypeDefinition(Position.expand(a, oldName.getPosition()), name, oldName);
		} else if (next.getToken().isType()) {
			Type type = parseType();

			return new TypeDefinition(Position.expand(a, type.getPosition()), name, type);
		}

		List<Field> fields = parseFieldList();
		Position b = null;
		if (fields.size() != 0) {
			b = Position
					.expand(fields.get(0).getPosition(), fields.get(fields.size() - 1)
							.getPosition());
		} else {
			b = current.getPosition();
		}

		return new TypeDefinition(Position.expand(a, b), name, fields);
	}

	private EnumDefinition parseEnumDefinition() {
		check(Token.KEYWORD_ENUM);
		Position a = current.getPosition();
		Identifier name = parseIdentifier();
		List<Identifier> values = parseIdentifierList();

		Position b = null;

		if (values.size() != 0) {
			b = Position
					.expand(values.get(0).getPosition(), values.get(values.size() - 1)
							.getPosition());
		} else {
			b = current.getPosition();
		}

		return new EnumDefinition(Position.expand(a, b), name, values);
	}

	private Field parseField() {
		Identifier name = parseIdentifier();
		Type type = parseType();

		return new Field(Position.expand(name.getPosition(), type.getPosition()), name, type);
	}

	private List<Field> parseFieldList() {
		check(Token.OPEN_BRACE);
		List<Field> fields = new ArrayList<Field>();

		while (match(Token.IDENTIFIER)) {
			fields.add(parseField());
			accept(Token.COMMA);
		}

		check(Token.CLOSE_BRACE);

		return fields;
	}

	private Type parseType() {
		Position a = current.getPosition();
		boolean isList = accept(Token.BRACKETS);

		if (match(Token.OPEN_BRACE)) {
			List<Field> fields = parseFieldList();
			Position b = null;

			if (fields.size() != 0) {
				b = Position
						.expand(fields.get(0).getPosition(), fields.get(fields.size() - 1)
								.getPosition());
			} else {
				b = current.getPosition();
			}

			return new Type(Position.expand(a, b), fields);
		}

		boolean isReference = accept(Token.ASTERISK);

		if (match(Token.IDENTIFIER)) {
			Identifier name = parseIdentifier();
			Position b = name.getPosition();

			return new Type(Position.expand(a, b), name, isList, isReference);
		} else {
			if (!next.getToken().isType()) {
				throw ParserException.expectedType(current.getPosition(), next.getToken());
			}

			accept(next.getToken());
			Position b = current.getPosition();

			return new Type(Position.expand(a, b), isList, isReference, current.getToken());
		}
	}

	private List<Identifier> parseIdentifierList() {
		check(Token.OPEN_BRACE);
		List<Identifier> identifiers = new ArrayList<Identifier>();

		while (match(Token.IDENTIFIER)) {
			identifiers.add(parseIdentifier());
			accept(Token.COMMA);
		}

		check(Token.CLOSE_BRACE);

		return identifiers;
	}

	private Identifier parseIdentifier() {
		check(Token.IDENTIFIER);

		return new Identifier(current.getPosition(), current.getLexeme());
	}

	private boolean match(Token token) {
		return next.getToken() == token;
	}

	private boolean match(Token... tokens) {
		for (int i = 0; i < tokens.length; i++) {
			if (match(tokens[i])) {
				return true;
			}
		}

		return false;
	}

	private void check(Token token) {
		if (match(token)) {
			current = next;
			next = lexer.next();
		} else {
			throw ParserException.unexpected(current.getPosition(), token, next.getToken());
		}
	}

	private void check(Token... tokens) {
		for (int i = 0; i < tokens.length; i++) {
			check(tokens[i]);
		}
	}

	private boolean accept(Token token) {
		if (match(token)) {
			current = next;
			next = lexer.next();

			return true;
		}

		return false;
	}

	private boolean accept(Token... tokens) {
		for (int i = 0; i < tokens.length; i++) {
			if (accept(tokens[i])) {
				return true;
			}
		}

		return false;
	}

}
